### DISCUSSION TITLE
[Suggestion] Add IP Blocker?

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#5772] [Ideas]
Any interest in including [IP Blocker](https://github.com/lat9/ip_blocker) in the core?

**Comment by torvista** [comment]
I needed this a couple of weeks ago for some russian IP (35k+ access attempts with odd chars in the URL).

**Comment by Pan2020** [comment]
I do not know why these Russian IP flooding so much daily (and worse that I can't filter them as they type in Russian and I could not filter properly).
Indeed, it will be great implementation to the core!

Hope this not related to 2020 stuffs...

**Comment by reetp** [comment]
Just implemented GeoIP blocking on my contact form using the Maxmind code. I just allow access from home nation as we've never had a genuine buyer from RU.....

No, it may not be perfect, but it's cut a lot of spam (I'm not a lover of captcha/google spyware). It's as blunt as the spammers :-)

Could easily be implemented elsewhere.

**Comment by scottcwilson** [comment]
Glad to hear you had a positive experience with Maxmind.  I wonder if your whole site can be Maxmind protected?  What's the fee?

**Comment by lankeeyankee** [comment]
This could be good. I block by country in WHM which has drastically cut down on spam and stupid stuff.

**Comment by reetp** [comment]
GeoLite DBs are free.

There are costs if you want better quality/more info but in my experience the basic one have always been OK

https://dev.maxmind.com/geoip/geoip2/geolite2/
https://www.maxmind.com/en/geoip2-databases
https://github.com/maxmind/MaxMind-DB-Reader-php

(I have absolutely no affiliation to Maxmind whatsoever - just used their stuff for a long time)

I would imagine it is pretty trivial to protect your entire site.

Effectively you do:

Clone out their code someplace - you'll need a cron or similar to autoupdate the DBs. I already use it for email blocking so I have that in place.

Then:

```
require_once 'MaxMind-DB-Reader-php/autoload.php';
use MaxMind\Db\Reader;
$databaseFile = '/usr/share/GeoIP/GeoLite2-Country.mmdb';
$reader = new Reader($databaseFile);
$IP_info = $reader->get($_SERVER['REMOTE_ADDR']);
$reader->close();

if ($IP_info['country']['iso_code'] == 'DE') {
// Show site
} else {
// Show some warning and no site
}
```

I'm no coder but you get the idea.....

One thing regarding the IP blocker, which then relates to this. There should probably be a way of whitelisting some stuff - either IPs, countries, or domains.

Neither method completely blocks all junk, but it may reduce it to manageable proportions.

Another fairly effective GeoIP method for those on Linux servers is the use of XT Tables addons with xt_geoip tables. It is a lot lighter than failban.

**Comment by scottcwilson** [comment]
Note that there are some IP tracking features already included in 1.5.8 (IP address used for account creation, for example), and this would be a nice addition.  Assigning to @lat9.

**Comment by lat9** [comment]
Since this suggestion was written &mdash; _4 years ago_ &mdash; there have been a couple of more advanced IP blocker plugins submitted:
- AbuseIPDB (https://www.zen-cart.com/downloads.php?do=file&id=1336)
- Access Blocker (https://www.zen-cart.com/downloads.php?do=file&id=2237)

Both of those plugins enable a site to block an access based on a cloud-based database, such that you can block by more than an IP address, which saves a site from having to update its blocked-ip-list every time a nefarious address hits the site.
